politeiad/backendv2/tstorebe/plugins/pi/cmds.go

```diff
@@ -242,25 +242,96 @@ func (p *piPlugin) cmdBillingStatusChanges(token []byte) (string, error) {
 	return string(reply), nil
 }
 
+// isFinalStatus determines whether the given proposal status is final and
+// not expected to change in the future.
+func isFinalStatus(status pi.PropStatusT) bool {
+	switch status {
+	case pi.PropStatusUnvettedAbandoned, pi.PropStatusUnvettedCensored,
+		pi.PropStatusAbandoned, pi.PropStatusCensored, pi.PropStatusApproved,
+		pi.PropStatusRejected:
+		return true
+	default:
+		return false
+	}
+}
+
+// needsBillingStatusChanges returns true if the given proposal status
+// is associated with approved proposal which needs only the latest billing
+// status metadata to detemine the proposal proposal status on runtime.
+func needsOnlyBillingStatusChanges(status pi.PropStatusT) bool {
+	switch status {
+	case pi.PropStatusActive, pi.PropStatusCompleted, pi.PropStatusClosed:
+		return true
+	default:
+		return false
+	}
+}
+
 // cmdSummary returns the pi summary of a proposal.
 func (p *piPlugin) cmdSummary(token []byte) (string, error) {
-	// Get an abridged version of the record. We only
-	// need the record metadata and the vote metadata.
-	r, err := p.record(backend.RecordRequest{
+	var (
+		r        *backend.Record
+		mdState  backend.StateT
+		mdStatus backend.StatusT
+
+		s          pi.PropStatusT
+		voteStatus ticketvote.VoteStatusT
+
+		voteMD *ticketvote.VoteMetadata
+		bscs   []pi.BillingStatusChange
+		err    error
+	)
+
+	// Check if any data associated with the token exists in the in-memory
+	// cache to avoid extra expensive full tlog tree reads.
+	tokenStr := hex.EncodeToString(token)
+	d := p.statuses.get(tokenStr)
+
+	// If no entry found in cache jump to fetch the record
+	if d == nil {
+		goto fetchrecord
+	}
+
+	// Cache entry found, determine proposal status using the cached data
+	switch {
+	// If the cached proposal status is final, jump to reply
+	case isFinalStatus(d.status):
+		s = d.status
+		goto reply
+
+	// If the cached proposal status needs latest billing status changes
+	// fetch them and determine the proposal status on runtime. This still
+	// avoids reading the full tlog tree.
+	case needsOnlyBillingStatusChanges(d.status):
+		bscs, err = p.billingStatusChanges(token)
+		if err != nil {
+			return "", err
+		}
+		s, err = proposalStatusApproved(nil, bscs)
```

Comment:
```
		s, err = proposalStatusApproved(nil, bscs)
		if err != nil {
			return "", err
		}
		// If runtime status different than cached status, cache the new status.
		if s != d.status {
			p.statuses.set(tokenStr, s)
		}
		goto reply
```

I'm not a fan of this.  You're adding in multiple code paths that determine the status and set the cache when it's possible to write this command in a way that only has a single path for determining the status and setting the cache.  Less logic paths means less complixity which is easier to reason about, write tests for, and maintain.  Notice that's how I described it in this comment.

https://github.com/decred/politeia/pull/1586#discussion_r757906449

If a function requires multiple logic paths then that's fine, but in this case, it doesn't.